File: chinonso_egeolu/project10/turing-machine.js
#!/usr/bin/env node

/**
 * Turing Machine Simulator - JavaScript Implementation
 * Supports both CLI and programmatic usage
 */

const readline = require('readline');

class TuringMachine {
    constructor(config) {
        this.states = config.states;
        this.alphabet = config.alphabet;
        this.transitions = config.transitions;
        this.startState = config.startState;
        this.acceptState = config.acceptState;
        this.rejectState = config.rejectState;
        this.blankSymbol = config.blankSymbol || '_';
        
        // Runtime state
        this.tape = new Map();
        this.headPosition = 0;
        this.currentState = this.startState;
        this.stepCount = 0;
        this.trace = [];
    }
    
    loadInput(input) {
        this.tape = new Map();
        for (let i = 0; i < input.length; i++) {
            this.tape.set(i, input[i]);
        }
        this.headPosition = 0;
        this.currentState = this.startState;
        this.stepCount = 0;
        this.trace = [];
    }
    
    readTape() {
        return this.tape.get(this.headPosition) || this.blankSymbol;
    }
    
    writeTape(symbol) {
        if (symbol === this.blankSymbol) {
            this.tape.delete(this.headPosition);
        } else {
            this.tape.set(this.headPosition, symbol);
        }
    }
    
    moveHead(direction) {
        if (direction === 'L') {
            this.headPosition--;
        } else if (direction === 'R') {
            this.headPosition++;
        }
    }
    
    getTapeString(window = 20) {
        if (this.tape.size === 0) {
            return `[${this.blankSymbol}]`;
        }
        
        const positions = Array.from(this.tape.keys());
        const minPos = Math.min(...positions, this.headPosition - window);
        const maxPos = Math.max(...positions, this.headPosition + window);
        
        let result = '';
        for (let i = minPos; i <= maxPos; i++) {
            const symbol = this.tape.get(i) || this.blankSymbol;
            if (i === this.headPosition) {
                result += `[${symbol}]`;
            } else {
                result += symbol;
            }
        }
        return result;
    }
    
    step() {
        const currentSymbol = this.readTape();
        
        // Record trace
        this.trace.push({
            step: this.stepCount,
            state: this.currentState,
            head: this.headPosition,
            symbol: currentSymbol,
            tape: this.getTapeString()
        });
        
        // Check halting states
        if (this.currentState === this.acceptState || 
            this.currentState === this.rejectState) {
            return false;
        }
        
        // Look up transition
        const key = `${this.currentState},${currentSymbol}`;
        if (!this.transitions[key]) {
            this.currentState = this.rejectState;
            return false;
        }
        
        const [newState, newSymbol, direction] = this.transitions[key];
        
        // Execute transition
        this.writeTape(newSymbol);
        this.currentState = newState;
        this.moveHead(direction);
        this.stepCount++;
        
        return true;
    }
    
    run(input, maxSteps = 1000) {
        this.loadInput(input);
        
        while (this.stepCount < maxSteps) {
            if (!this.step()) break;
        }
        
        return {
            accepted: this.currentState === this.acceptState,
            trace: this.trace,
            steps: this.stepCount,
            finalState: this.currentState
        };
    }
}

// =============================================================================
// PRE-BUILT MACHINES
// =============================================================================

function createBinaryPalindromeChecker() {
    return new TuringMachine({
        states: ['q0', 'q1', 'q2', 'q3', 'q4', 'q5', 'accept', 'reject'],
        alphabet: ['0', '1'],
        startState: 'q0',
        acceptState: 'accept',
        rejectState: 'reject',
        transitions: {
            'q0,0': ['q1', 'X', 'R'],
            'q0,1': ['q2', 'X', 'R'],
            'q0,X': ['q0', 'X', 'R'],
            'q0,_': ['accept', '_', 'S'],
            
            'q1,0': ['q1', '0', 'R'],
            'q1,1': ['q1', '1', 'R'],
            'q1,X': ['q1', 'X', 'R'],
            'q1,_': ['q3', '_', 'L'],
            
            'q3,0': ['q4', 'X', 'L'],
            'q3,X': ['q0', 'X', 'R'],
            
            'q2,0': ['q2', '0', 'R'],
            'q2,1': ['q2', '1', 'R'],
            'q2,X': ['q2', 'X', 'R'],
            'q2,_': ['q5', '_', 'L'],
            
            'q5,1': ['q4', 'X', 'L'],
            'q5,X': ['q0', 'X', 'R'],
            
            'q4,0': ['q4', '0', 'L'],
            'q4,1': ['q4', '1', 'L'],
            'q4,X': ['q4', 'X', 'L'],
            'q4,_': ['q0', '_', 'R']
        }
    });
}

function createBalancedParenthesesChecker() {
    return new TuringMachine({
        states: ['q0', 'q1', 'q2', 'accept', 'reject'],
        alphabet: ['(', ')'],
        startState: 'q0',
        acceptState: 'accept',
        rejectState: 'reject',
        transitions: {
            'q0,(': ['q1', 'X', 'R'],
            'q0,X': ['q0', 'X', 'R'],
            'q0,_': ['accept', '_', 'S'],
            
            'q1,(': ['q1', '(', 'R'],
            'q1,)': ['q2', 'X', 'L'],
            'q1,X': ['q1', 'X', 'R'],
            
            'q2,(': ['q2', '(', 'L'],
            'q2,X': ['q2', 'X', 'L'],
            'q2,_': ['q0', '_', 'R']
        }
    });
}

function createAnBnChecker() {
    return new TuringMachine({
        states: ['q0', 'q1', 'q2', 'accept', 'reject'],
        alphabet: ['a', 'b'],
        startState: 'q0',
        acceptState: 'accept',
        rejectState: 'reject',
        transitions: {
            'q0,a': ['q1', 'X', 'R'],
            'q0,Y': ['q0', 'Y', 'R'],
            'q0,_': ['accept', '_', 'S'],
            
            'q1,a': ['q1', 'a', 'R'],
            'q1,Y': ['q1', 'Y', 'R'],
            'q1,b': ['q2', 'Y', 'L'],
            
            'q2,a': ['q2', 'a', 'L'],
            'q2,Y': ['q2', 'Y', 'L'],
            'q2,X': ['q2', 'X', 'L'],
            'q2,_': ['q0', '_', 'R']
        }
    });
}

// =============================================================================
// CLI INTERFACE
// =============================================================================

const machines = {
    '1': { name: 'Binary Palindrome Checker', creator: createBinaryPalindromeChecker },
    '2': { name: 'Balanced Parentheses Checker', creator: createBalancedParenthesesChecker },
    '3': { name: 'a^n b^n Language', creator: createAnBnChecker }
};

function printMenu() {
    console.log('\n' + '='.repeat(70));
    console.log('TURING MACHINE SIMULATOR');
    console.log('='.repeat(70));
    console.log('\nAvailable Machines:');
    console.log('  1. Binary Palindrome Checker (e.g., "1001", "0110")');
    console.log('  2. Balanced Parentheses Checker (e.g., "(())", "()()")');
    console.log('  3. a^n b^n Language (e.g., "aabb", "aaabbb")');
    console.log('  4. Exit');
    console.log();
}

function printTrace(trace, verbose = true) {
    if (verbose) {
        console.log('\n' + 'Step'.padEnd(6) + 'State'.padEnd(10) + 'Head'.padEnd(6) + 
                    'Symbol'.padEnd(8) + 'Tape');
        console.log('-'.repeat(70));
        trace.forEach(entry => {
            console.log(
                String(entry.step).padEnd(6) +
                entry.state.padEnd(10) +
                String(entry.head).padEnd(6) +
                entry.symbol.padEnd(8) +
                entry.tape
            );
        });
    } else {
        console.log(`\nShowing first and last 5 steps of ${trace.length} total:`);
        console.log('\n' + 'Step'.padEnd(6) + 'State'.padEnd(10) + 'Symbol'.padEnd(8));
        console.log('-'.repeat(30));
        
        trace.slice(0, 5).forEach(entry => {
            console.log(
                String(entry.step).padEnd(6) +
                entry.state.padEnd(10) +
                entry.symbol.padEnd(8)
            );
        });
        
        if (trace.length > 10) {
            console.log('  ...');
        }
        
        trace.slice(-5).forEach(entry => {
            console.log(
                String(entry.step).padEnd(6) +
                entry.state.padEnd(10) +
                entry.symbol.padEnd(8)
            );
        });
    }
    console.log();
}

async function runInteractive() {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
    
    const question = (query) => new Promise(resolve => rl.question(query, resolve));
    
    while (true) {
        printMenu();
        const choice = await question('Select a machine (1-4): ');
        
        if (choice === '4') {
            console.log('\nGoodbye!');
            rl.close();
            break;
        }
        
        if (!machines[choice]) {
            console.log('\nInvalid choice. Please try again.');
            continue;
        }
        
        const { name, creator } = machines[choice];
        console.log('\n' + '='.repeat(70));
        console.log(`Selected: ${name}`);
        console.log('='.repeat(70));
        
        const input = await question('\nEnter input string (or "back" to return): ');
        
        if (input.toLowerCase() === 'back') {
            continue;
        }
        
        console.log('\n' + '='.repeat(70));
        console.log(`Running Turing Machine on input: '${input}'`);
        console.log('='.repeat(70));
        
        const tm = creator();
        const result = tm.run(input);
        
        const verbose = result.trace.length <= 50;
        printTrace(result.trace, verbose);
        
        console.log('='.repeat(70));
        if (result.accepted) {
            console.log(`✓ ACCEPTED - Input '${input}' is in the language`);
        } else {
            console.log(`✗ REJECTED - Input '${input}' is NOT in the language`);
        }
        console.log(`Total steps: ${result.steps}`);
        console.log(`Final state: ${result.finalState}`);
        console.log('='.repeat(70));
        
        await question('\nPress Enter to continue...');
    }
}

function runTests() {
    console.log('Running automated tests...\n');
    
    // Test 1: Palindrome
    console.log('Test 1: Binary Palindrome');
    let tm = createBinaryPalindromeChecker();
    ['101', '1001', '110', ''].forEach(test => {
        const result = tm.run(test);
        console.log(`  '${test}': ${result.accepted ? 'PASS' : 'FAIL'}`);
    });
    
    // Test 2: Parentheses
    console.log('\nTest 2: Balanced Parentheses');
    tm = createBalancedParenthesesChecker();
    ['()', '(())', '(()', ''].forEach(test => {
        const result = tm.run(test);
        console.log(`  '${test}': ${result.accepted ? 'PASS' : 'FAIL'}`);
    });
    
    // Test 3: a^n b^n
    console.log('\nTest 3: a^n b^n');
    tm = createAnBnChecker();
    ['ab', 'aabb', 'aaabbb', 'aab', ''].forEach(test => {
        const result = tm.run(test);
        console.log(`  '${test}': ${result.accepted ? 'PASS' : 'FAIL'}`);
    });
}

// Main entry point
if (require.main === module) {
    if (process.argv[2] === '--test') {
        runTests();
    } else {
        runInteractive();
    }
}

// Export for web server
module.exports = {
    TuringMachine,
    createBinaryPalindromeChecker,
    createBalancedParenthesesChecker,
    createAnBnChecker
};

File: raegan_green/project10/turing_machine.py
#!/usr/bin/env python3
"""
Turing Machine Simulator
A comprehensive implementation with state tracing and multiple built-in machines.
"""

import sys
from typing import List, Tuple, Optional, Dict
from enum import Enum


class Direction(Enum):
    """Tape head movement direction."""
    LEFT = -1
    RIGHT = 1
    STAY = 0


class TuringMachine:
    """
    A Turing Machine simulator with full state tracing.
    
    The machine operates on a tape with cells containing symbols.
    State transitions are defined by (current_state, read_symbol) -> (new_state, write_symbol, direction)
    """
    
    def __init__(self, 
                 states: set,
                 input_alphabet: set,
                 tape_alphabet: set,
                 transitions: Dict[Tuple[str, str], Tuple[str, str, Direction]],
                 start_state: str,
                 accept_state: str,
                 reject_state: str,
                 blank_symbol: str = '_'):
        """
        Initialize the Turing Machine.
        
        Args:
            states: Set of state names
            input_alphabet: Set of input symbols
            tape_alphabet: Set of tape symbols (includes input_alphabet + blank + work symbols)
            transitions: Dictionary mapping (state, symbol) -> (new_state, write_symbol, direction)
            start_state: Initial state
            accept_state: Accepting state
            reject_state: Rejecting state
            blank_symbol: Symbol representing blank cells
        """
        self.states = states
        self.input_alphabet = input_alphabet
        self.tape_alphabet = tape_alphabet
        self.transitions = transitions
        self.start_state = start_state
        self.accept_state = accept_state
        self.reject_state = reject_state
        self.blank_symbol = blank_symbol
        
        # Execution state
        self.tape: List[str] = []
        self.head_position: int = 0
        self.current_state: str = start_state
        self.step_count: int = 0
        self.trace: List[Dict] = []
        self.max_steps: int = 10000  # Prevent infinite loops
    
    def initialize_tape(self, input_string: str):
        """Initialize the tape with input string."""
        self.tape = list(input_string) if input_string else [self.blank_symbol]
        self.head_position = 0
        self.current_state = self.start_state
        self.step_count = 0
        self.trace = []
        
        # Record initial configuration
        self._record_trace("INIT")
    
    def _record_trace(self, action: str = "STEP"):
        """Record current configuration for trace."""
        # Get tape representation with context
        left_context = max(0, self.head_position - 5)
        right_context = min(len(self.tape), self.head_position + 6)
        
        # Ensure tape has content
        if not self.tape:
            self.tape = [self.blank_symbol]
        
        tape_view = ''.join(self.tape[left_context:right_context])
        head_offset = self.head_position - left_context
        
        self.trace.append({
            'step': self.step_count,
            'action': action,
            'state': self.current_state,
            'head_position': self.head_position,
            'tape_view': tape_view,
            'head_offset': head_offset,
            'read_symbol': self.tape[self.head_position] if 0 <= self.head_position < len(self.tape) else self.blank_symbol
        })
    
    def step(self) -> bool:
        """
        Execute one step of the Turing Machine.
        
        Returns:
            True if machine should continue, False if halted (accept/reject)
        """
        # Extend tape if needed
        if self.head_position < 0:
            self.tape.insert(0, self.blank_symbol)
            self.head_position = 0
        elif self.head_position >= len(self.tape):
            self.tape.append(self.blank_symbol)
        
        # Read current symbol
        current_symbol = self.tape[self.head_position]
        
        # Check if we're in a halting state
        if self.current_state == self.accept_state:
            self._record_trace("ACCEPT")
            return False
        if self.current_state == self.reject_state:
            self._record_trace("REJECT")
            return False
        
        # Get transition
        transition_key = (self.current_state, current_symbol)
        if transition_key not in self.transitions:
            # No transition defined - implicit reject
            self.current_state = self.reject_state
            self._record_trace("REJECT (no transition)")
            return False
        
        new_state, write_symbol, direction = self.transitions[transition_key]
        
        # Execute transition
        self.tape[self.head_position] = write_symbol
        self.current_state = new_state
        self.head_position += direction.value
        self.step_count += 1
        
        # Record trace
        self._record_trace()
        
        # Check for infinite loop
        if self.step_count >= self.max_steps:
            self.current_state = self.reject_state
            self._record_trace("REJECT (max steps exceeded)")
            return False
        
        return True
    
    def run(self, input_string: str) -> Tuple[bool, str]:
        """
        Run the Turing Machine on input string.
        
        Args:
            input_string: Input to process
            
        Returns:
            Tuple of (accepted: bool, result_message: str)
        """
        self.initialize_tape(input_string)
        
        while self.step():
            pass
        
        accepted = self.current_state == self.accept_state
        
        if accepted:
            result = f"ACCEPTED after {self.step_count} steps"
        else:
            result = f"REJECTED after {self.step_count} steps"
        
        return accepted, result
    
    def print_trace(self, verbose: bool = True):
        """Print execution trace."""
        print("\n" + "="*70)
        print("EXECUTION TRACE")
        print("="*70)
        
        for entry in self.trace:
            if verbose or entry['action'] in ['INIT', 'ACCEPT', 'REJECT']:
                step_str = f"Step {entry['step']:4d}" if entry['step'] > 0 else "Initial"
                print(f"\n{step_str} | State: {entry['state']:15s} | Action: {entry['action']}")
                
                if verbose:
                    # Print tape with head indicator
                    print(f"  Tape:  {entry['tape_view']}")
                    print(f"  Head:  {' ' * entry['head_offset']}^")
                    print(f"  Read:  {entry['read_symbol']}")
        
        print("\n" + "="*70)


# ==============================================================================
# Pre-defined Turing Machines
# ==============================================================================

def create_binary_palindrome_checker() -> TuringMachine:
    """
    Creates a Turing Machine that checks if a binary string is a palindrome.
    
    Algorithm:
    1. Mark leftmost unmarked symbol, remember it
    2. Move right to find rightmost unmarked symbol
    3. Check if they match
    4. If match, repeat; if all checked, accept; if mismatch, reject
    
    Complexity: O(n²) time, recognizes language {w | w = w^R, w ∈ {0,1}*}
    """
    states = {'q0', 'q1', 'q2', 'q3', 'q4', 'q5', 'q6', 'q_accept', 'q_reject'}
    input_alphabet = {'0', '1'}
    tape_alphabet = {'0', '1', 'X', 'Y', '_'}
    
    transitions = {
        # Start: mark left symbol and remember it
        ('q0', '0'): ('q1', 'X', Direction.RIGHT),  # Remember 0
        ('q0', '1'): ('q2', 'X', Direction.RIGHT),  # Remember 1
        ('q0', 'X'): ('q0', 'X', Direction.RIGHT),  # Skip marked
        ('q0', 'Y'): ('q0', 'Y', Direction.RIGHT),  # Skip marked
        ('q0', '_'): ('q_accept', '_', Direction.STAY),  # Empty or all checked
        
        # q1: remembered 0, move to rightmost
        ('q1', '0'): ('q1', '0', Direction.RIGHT),
        ('q1', '1'): ('q1', '1', Direction.RIGHT),
        ('q1', 'Y'): ('q1', 'Y', Direction.RIGHT),
        ('q1', '_'): ('q3', '_', Direction.LEFT),  # Found end
        
        # q2: remembered 1, move to rightmost
        ('q2', '0'): ('q2', '0', Direction.RIGHT),
        ('q2', '1'): ('q2', '1', Direction.RIGHT),
        ('q2', 'Y'): ('q2', 'Y', Direction.RIGHT),
        ('q2', '_'): ('q4', '_', Direction.LEFT),  # Found end
        
        # q3: check if rightmost is 0
        ('q3', '0'): ('q5', 'Y', Direction.LEFT),  # Match! Mark and go back
        ('q3', '1'): ('q_reject', '1', Direction.STAY),  # Mismatch
        ('q3', 'Y'): ('q3', 'Y', Direction.LEFT),  # Skip marked
        ('q3', 'X'): ('q_accept', 'X', Direction.STAY),  # All checked (single or even)
        
        # q4: check if rightmost is 1
        ('q4', '1'): ('q5', 'Y', Direction.LEFT),  # Match! Mark and go back
        ('q4', '0'): ('q_reject', '0', Direction.STAY),  # Mismatch
        ('q4', 'Y'): ('q4', 'Y', Direction.LEFT),  # Skip marked
        ('q4', 'X'): ('q_accept', 'X', Direction.STAY),  # All checked
        
        # q5: move back to start
        ('q5', '0'): ('q5', '0', Direction.LEFT),
        ('q5', '1'): ('q5', '1', Direction.LEFT),
        ('q5', 'X'): ('q5', 'X', Direction.LEFT),
        ('q5', 'Y'): ('q5', 'Y', Direction.LEFT),
        ('q5', '_'): ('q6', '_', Direction.RIGHT),  # Reached start
        
        # q6: return to start state
        ('q6', 'X'): ('q0', 'X', Direction.RIGHT),
        ('q6', 'Y'): ('q0', 'Y', Direction.RIGHT),
    }
    
    return TuringMachine(
        states=states,
        input_alphabet=input_alphabet,
        tape_alphabet=tape_alphabet,
        transitions=transitions,
        start_state='q0',
        accept_state='q_accept',
        reject_state='q_reject',
        blank_symbol='_'
    )


def create_binary_increment() -> TuringMachine:
    """
    Creates a Turing Machine that increments a binary number by 1.
    
    Algorithm:
    1. Move to rightmost digit
    2. Add 1 (handle carry)
    3. Propagate carry left if needed
    
    Complexity: O(n) time, computes f(w) = binary(decimal(w) + 1)
    """
    states = {'q0', 'q1', 'q2', 'q_accept', 'q_reject'}
    input_alphabet = {'0', '1'}
    tape_alphabet = {'0', '1', '_'}
    
    transitions = {
        # Move to rightmost digit
        ('q0', '0'): ('q0', '0', Direction.RIGHT),
        ('q0', '1'): ('q0', '1', Direction.RIGHT),
        ('q0', '_'): ('q1', '_', Direction.LEFT),  # Found end
        
        # Add 1 to rightmost
        ('q1', '0'): ('q_accept', '1', Direction.STAY),  # 0+1=1, done
        ('q1', '1'): ('q2', '0', Direction.LEFT),  # 1+1=0, carry
        ('q1', '_'): ('q_accept', '1', Direction.STAY),  # Empty input -> 1
        
        # Propagate carry
        ('q2', '0'): ('q_accept', '1', Direction.STAY),  # 0+carry=1, done
        ('q2', '1'): ('q2', '0', Direction.LEFT),  # 1+carry=0, continue carry
        ('q2', '_'): ('q_accept', '1', Direction.STAY),  # Overflow, prepend 1
    }
    
    return TuringMachine(
        states=states,
        input_alphabet=input_alphabet,
        tape_alphabet=tape_alphabet,
        transitions=transitions,
        start_state='q0',
        accept_state='q_accept',
        reject_state='q_reject',
        blank_symbol='_'
    )


def create_anbn_recognizer() -> TuringMachine:
    """
    Creates a Turing Machine that recognizes the language {a^n b^n | n >= 0}.
    
    Algorithm:
    1. Mark one 'a' and one 'b'
    2. Repeat until all symbols are marked
    3. Accept if equal count, reject otherwise
    
    Complexity: O(n²) time, classic context-free language
    """
    states = {'q0', 'q1', 'q2', 'q3', 'q4', 'q_accept', 'q_reject'}
    input_alphabet = {'a', 'b'}
    tape_alphabet = {'a', 'b', 'X', 'Y', '_'}
    
    transitions = {
        # Start: find first unmarked 'a'
        ('q0', 'a'): ('q1', 'X', Direction.RIGHT),  # Mark 'a'
        ('q0', 'X'): ('q0', 'X', Direction.RIGHT),  # Skip marked 'a'
        ('q0', 'Y'): ('q0', 'Y', Direction.RIGHT),  # Skip marked 'b'
        ('q0', 'b'): ('q_reject', 'b', Direction.STAY),  # b before a
        ('q0', '_'): ('q_accept', '_', Direction.STAY),  # Empty or all matched
        
        # q1: skip remaining 'a's and X's, find first 'b'
        ('q1', 'a'): ('q1', 'a', Direction.RIGHT),
        ('q1', 'X'): ('q1', 'X', Direction.RIGHT),
        ('q1', 'b'): ('q2', 'Y', Direction.LEFT),  # Mark corresponding 'b'
        ('q1', 'Y'): ('q1', 'Y', Direction.RIGHT),  # Skip marked 'b'
        ('q1', '_'): ('q_reject', '_', Direction.STAY),  # No matching 'b'
        
        # q2: return to start
        ('q2', 'a'): ('q2', 'a', Direction.LEFT),
        ('q2', 'b'): ('q2', 'b', Direction.LEFT),
        ('q2', 'X'): ('q2', 'X', Direction.LEFT),
        ('q2', 'Y'): ('q2', 'Y', Direction.LEFT),
        ('q2', '_'): ('q3', '_', Direction.RIGHT),
        
        # q3: move to next unmarked 'a'
        ('q3', 'X'): ('q0', 'X', Direction.RIGHT),
        ('q3', 'Y'): ('q4', 'Y', Direction.RIGHT),  # Check if all done
        
        # q4: verify all b's are marked
        ('q4', 'Y'): ('q4', 'Y', Direction.RIGHT),
        ('q4', '_'): ('q_accept', '_', Direction.STAY),
        ('q4', 'b'): ('q_reject', 'b', Direction.STAY),  # Extra b
    }
    
    return TuringMachine(
        states=states,
        input_alphabet=input_alphabet,
        tape_alphabet=tape_alphabet,
        transitions=transitions,
        start_state='q0',
        accept_state='q_accept',
        reject_state='q_reject',
        blank_symbol='_'
    )


# ==============================================================================
# Main Program
# ==============================================================================

def print_header():
    """Print program header."""
    print("\n" + "="*70)
    print(" " * 20 + "TURING MACHINE SIMULATOR")
    print("="*70)
    print("\nA comprehensive implementation with multiple built-in machines")
    print("and full state trace visualization.\n")


def print_machine_info(name: str, description: str, examples: List[Tuple[str, bool]]):
    """Print information about a Turing Machine."""
    print("\n" + "-"*70)
    print(f"Machine: {name}")
    print("-"*70)
    print(f"Description: {description}")
    print("\nExample inputs:")
    for input_str, should_accept in examples:
        status = "✓ Accept" if should_accept else "✗ Reject"
        display_input = input_str if input_str else "(empty)"
        print(f"  '{display_input}' -> {status}")
    print()


def main():
    """Main program entry point."""
    print_header()
    
    # Define available machines
    machines = {
        '1': {
            'name': 'Binary Palindrome Checker',
            'description': 'Checks if a binary string reads the same forwards and backwards',
            'creator': create_binary_palindrome_checker,
            'examples': [
                ('', True),
                ('0', True),
                ('1', True),
                ('11', True),
                ('101', True),
                ('0110', False),
                ('10101', True),
                ('110011', True),
                ('111000', False),
            ]
        },
        '2': {
            'name': 'Binary Increment',
            'description': 'Adds 1 to a binary number',
            'creator': create_binary_increment,
            'examples': [
                ('', True),  # -> 1
                ('0', True),  # -> 1
                ('1', True),  # -> 10
                ('10', True),  # -> 11
                ('11', True),  # -> 100
                ('101', True),  # -> 110
                ('111', True),  # -> 1000
            ]
        },
        '3': {
            'name': 'a^n b^n Recognizer',
            'description': 'Recognizes strings with equal number of a\'s followed by equal b\'s',
            'creator': create_anbn_recognizer,
            'examples': [
                ('', True),
                ('ab', True),
                ('aabb', True),
                ('aaabbb', True),
                ('aaaabbbb', True),
                ('a', False),
                ('b', False),
                ('aab', False),
                ('abb', False),
                ('ba', False),
                ('aabba', False),
            ]
        }
    }
    
    # Get machine selection
    if len(sys.argv) > 1 and sys.argv[1] in machines:
        choice = sys.argv[1]
    else:
        print("Available Turing Machines:")
        for key, machine in machines.items():
            print(f"  {key}. {machine['name']}")
        print("\nUsage: python turing_machine.py <machine_number> [input_string]")
        print("Example: python turing_machine.py 1 101")
        print("\nOr run interactively:")
        print("  python turing_machine.py")
        print()
        
        choice = input("Select machine (1-3): ").strip()
        if choice not in machines:
            print(f"Invalid choice: {choice}")
            return 1
    
    # Get input string
    if len(sys.argv) > 2:
        input_string = sys.argv[2]
        interactive = False
    else:
        interactive = True
    
    # Create and display machine info
    machine_config = machines[choice]
    print_machine_info(
        machine_config['name'],
        machine_config['description'],
        machine_config['examples']
    )
    
    tm = machine_config['creator']()
    
    # Process input(s)
    if interactive:
        print("Enter input strings (one per line, 'quit' to exit):")
        print("Tip: Use '-' for empty string\n")
        
        while True:
            try:
                user_input = input("Input> ").strip()
                if user_input.lower() == 'quit':
                    break
                
                if user_input == '-':
                    user_input = ''
                
                print(f"\nProcessing: '{user_input if user_input else '(empty)'}'")
                print("-"*70)
                
                accepted, message = tm.run(user_input)
                print(f"\nResult: {message}")
                tm.print_trace(verbose=True)
                
                print("\n")
                
            except KeyboardInterrupt:
                print("\n\nExiting...")
                break
            except EOFError:
                break
    else:
        print(f"Processing: '{input_string if input_string else '(empty)'}'")
        print("-"*70)
        
        accepted, message = tm.run(input_string)
        print(f"\nResult: {message}")
        tm.print_trace(verbose=True)
    
    return 0


if __name__ == "__main__":
    sys.exit(main())
    